Feature/open planner dev

hatem-darweesh opened the pull request:
## Status
** DEVELOPMENT**

## Description
Add OpenPlanner new Modifications and Tools to Autoware develop branch.
- OpenPlanner - Global Planner 
- OpenPlanner - Local Planner 
- OpenPlanner - Utilities
- OpenPlanner - Simulator 
- lidar_kf_contour_track 
Remove Old OpenPlanner (dp_planner, way_planner) 
Add readme for each module 

**For More Details about the updated OpenPlanner, interface description and how to test **
[OpenPlanner Nodes.pdf](https://github.com/CPFL/Autoware/files/2300277/OpenPlanner.Nodes.pdf)


## Steps to Test or Reproduce
 * Should produce same behaviour as previous (way_planner, dp_planner). 
 * Should show same output as [Demo Video](https://youtu.be/BS5nLtBsXPE)
 * Basic Functions To Test:

1. OpenPlanner - Global planner should load map successfully and generate global path from start to goal points specified by rviz or from .csv file. 
1. OpenPlanner - Local Planner should generate local trajectories along the global path, follow other simulated vehicles, avoid obstacles, stop for stop lines, and stop in traffic lights if light is red. 
1. OpenPlanner - Utilities should be able to play rosbag file including localization, pause, play backwards, and  publish TF for ndt_pose
1. OpenPlanner - Simulator should simulate a vehicle moving with specific speed from start to goal position, simulate simple intersection traffic lights, and simulate perception by generating cloud_clusters for simulate vehicles. 
1. lidar_kf_contour_track should track detected objects, find heading and velocity for objects detected by lidar_euclidean_cluster or op_perception_simulator.  

Follow README files and [Demo Video](https://youtu.be/BS5nLtBsXPE)

hatem-darweesh committed:
Update OpenPlanner libraries (op_planner, op_utitity, op_ros_helpers)
Update ring ground filter with latest implementation
Update lidar_kf_contour_track with latest implementation
Add op_utilities nodes (op_bag_player, op_data_logger, op_pose2tf)
Modify autoware_msgs for OpenPlanner use (CloudCluster, DetectedObject, lane, waypoint)
Update UI computing.yaml for the new nodes and modifies parameters
Update UI sensing.yaml for updated ring_ground_filter params

hatem-darweesh committed:
Update op_utility files for csv files loading
Update MappingHelpers with latest modifications
Update PlanningHelpers with latest modifications
add op_common_param node, for setting OpenPlanner parameter for all related nodes such as lidar_kf_contour_track
Improve tracking by including size different in association function
Update way_planner, dp_planner for compatibility with new Mapping Modifications, Map format is backward compatible

hatem-darweesh committed:
Fix Vector Map parser problem, tested with three different maps
Fix Global Planning function for the new map modification
Add OpenPlanner Simulator for perception, traffic lights, cars
Add OpenPlanner new version to replace wp_planner and dp_planner
Remove unnecessary files from OpenPlanner libraries
Test Global and Local planning
Test Tracking node (kf_contour_track)
Test Simulation Nodes
Test Utility Nodes

hatem-darweesh committed:
Modify Map loading for OpenPlanner, now it reads from Autoware vector map messages, old behavior still works but from launch file only.
Delete way_planner, dp_planner from UI, but they still accessible from roslaunch.

hatem-darweesh committed:
Test Simulated Vehicles
Fix Simulated Vehicle Initialization
Test Following
Test Obstacle Avoidance
Add Visualization information to rviz config file open_planner.rviz

hatem-darweesh committed:
Add ReadMe to lidar_kf_contour_track node

hatem-darweesh committed:
Update README.md

hatem-darweesh committed:
Add README files for OpenPlanner packages

hatem-darweesh committed:
Merge branch 'feature/OpenPlanner-dev' of https://github.com/CPFL/Autoware into feature/OpenPlanner-dev

hatem-darweesh committed:
Update README.md

hatem-darweesh committed:
Update README.md

hatem-darweesh committed:
Merge branch 'develop' into feature/OpenPlanner-dev

hatem-darweesh committed:
Fix Indigo build issues

Commit:
fix dependencies to make industrial_ci test pass

Commit:
Merge branch 'develop' into feature/OpenPlanner-dev

kitsukawa commented on ros/src/util/packages/runtime_manager/scripts/computing.yaml:
```diff
@@ -530,35 +530,106 @@ subs :
+        desc : Simulate traffic Signs status for intersections included in the map
+        cmd  : roslaunch op_simulation_package op_signs_simulator.launch
+        param: op_signs_simulator
+      - name : op_car_simulator_1
```
@hatem-darweesh There are 5 checkboxes for op_car_simulator. This should be only one here.

kfunaoka replied:
@hatem-darweesh How about @kitsukawa's comment?

hatem-darweesh replied:
@kitsukawa @kfunaoka Only 1 check box for simulated car remaining, I will add multi car simulation option in "Simulation" tab later.

kfunaoka commented on ros/src/computing/perception/detection/lidar_tracker/packages/lidar_kf_contour_track/CMakeLists.txt:
```diff
@@ -8,10 +8,10 @@ find_package(catkin REQUIRED COMPONENTS
   )
 
-
-set(CMAKE_CXX_FLAGS "-O2 -Wall ${CMAKE_CXX_FLAGS}")
```
Is this deletion intended?

hatem-darweesh replied:
I saw a PR which adds common compiler flags to all cmake files "autoware_build_flags". so I deleted the special ones

kfunaoka replied:
Thanks! I should accomplish autoware_build_flags because it sets only -std=c++11 now.

hatem-darweesh replied:
The other flags are needed only locally for debugging and performance test so I don't want to push them to the server

kfunaoka commented on ros/src/computing/perception/detection/lidar_tracker/packages/lidar_kf_contour_track/include/PolygonGenerator.h:
```diff
@@ -0,0 +1,95 @@
+
```
Is it possible to distribute this file under BSD 3-Clause "New" or "Revised" License?
If yes, please insert the license terms.

hatem-darweesh replied:
can you give me an example to follow.

kfunaoka replied:
This is the license.
https://github.com/CPFL/Autoware/blob/master/LICENSE

An example is here. The year 2015 in it should be replaced with 2018 this year.
https://github.com/CPFL/Autoware/blob/develop/ros/src/computing/perception/detection/vision_detector/packages/vision_ssd_detect/src/vision_ssd_detect.cpp

kfunaoka commented on ros/src/computing/perception/detection/lidar_tracker/packages/lidar_kf_contour_track/include/PolygonGenerator.h:
```diff
@@ -0,0 +1,95 @@
+	PlannerHNS::WayPoint max_from_center;
+	bool bFirst;
+
+	QuarterView(const int& min_a, const int& max_a, const int& index)
```
It is better to use int or const int instead of const int&.

hatem-darweesh replied:
it is less confusing to use one parameter format for both objects and standard types. 
and it will be really dangerous and will take plenty of time and testing to change it through the whole OpenPlanner library.

kfunaoka replied:
Thanks. I think so too in the point of view where it takes much time and dangerous. I think it is better to be as is in this pull request.

But using const int & instead of int simply incurs piled-up overhead. I think C++ programmers do not confuse about it because they should know C and C++ basically employs pass by value.

kfunaoka commented on ros/src/computing/perception/detection/lidar_tracker/packages/lidar_kf_contour_track/include/PolygonGenerator.h:
```diff
@@ -0,0 +1,95 @@
+
+	bool UpdateQuarterView(const PlannerHNS::WayPoint& v)
+	{
+		if(v.pos.a <= min_ang || v.pos.a > max_ang)
```
It seems that clang-format was not used. Did you use clang-format? If no, please use clang-format to all the files you modified. (https://github.com/CPFL/Autoware/wiki/Contribution-Rules#ros-c-coding)

hatem-darweesh replied:
I will apply it later with different PR, after merging this one.

kfunaoka replied:
@hatem-darweesh Your choice is the best method.

kfunaoka commented on ros/src/computing/perception/detection/lidar_tracker/packages/lidar_kf_contour_track/include/SimpleTracker.h:
```diff
@@ -0,0 +1,440 @@
+namespace ContourTrackerNS
+{
+
+#define DEBUG_TRACKER 0
```
It's better to use constexpr instead of #define. (https://en.cppreference.com/w/cpp/language/constexpr)

hatem-darweesh replied:
Can we add this as ToDo, I need to check for each #define whether it is a compiler directive or a constant. so it will take long time.

kfunaoka replied:
@hatem-darweesh NP :)

kfunaoka commented on ros/src/computing/planning/common/lib/openplanner/op_planner/include/op_planner/BehaviorPrediction.h:
```diff
@@ -0,0 +1,882 @@
+		}
+
+		int nEqualities = 0;
+		for(unsigned int i=0; i < _ids.size(); i++)
```
It's better to use range-for if index is not required. (https://en.cppreference.com/w/cpp/language/range-for)

hatem-darweesh replied:
I see you prefer C++11 coding style,
This code is written in C and old C++ style. it is more readable for me, and other old-school programmers. if anything thing producing bugs or will affect the performance I will gladly change it.

kfunaoka replied:
@hatem-darweesh NP :)

I’ve written my opinion below. If you like range-based for, please use it in your future work.

- I’m not interested in C++11 coding style. I’m an old person studying old C++ before C++98 standardization. I think some of C++11 features such as std::move should not be used by most programmers.
- I’m interested in the tools that improve our work. Not different from ROS, Python, or Deep-Learning. Range-based for is just a tool we got.
- Why C++11 introduces range-based for is that it is a best practice in software engineering. Range-based for abstracts loops higher than index-based for without losing performance. Because we need not care index information in range-based for, the number/amount of characters/information we type, look, and consider in development and review can be reduced.
- Because range-based for is a best practice in software engineering, it is employed by other programming languages. Examples are Python https://docs.python.org/3/tutorial/controlflow.html#for-statements and Ruby https://ruby-doc.org/core-2.5.0/Array.html#class-Array-label-Iterating+over+Arrays. Of course, index-based for can be used in them, but it is not a good idea to use index-based for where index information is not required.

hatem-darweesh replied:
I agree with you. I will try to adopt these styles when I refactor the code.

kfunaoka commented on ros/src/computing/planning/common/lib/openplanner/op_planner/include/op_planner/BehaviorPrediction.h:
```diff
@@ -0,0 +1,882 @@
+		}
+	};
+
+	void DeleteFromList(std::vector<TrajectoryTracker*>& delete_me_track, TrajectoryTracker* track)
```
track should be const TrajectoryTracker *

hatem-darweesh replied:
Done

kfunaoka commented on ros/src/computing/planning/common/lib/openplanner/op_planner/include/op_planner/BehaviorPrediction.h:
```diff
@@ -0,0 +1,882 @@
+			f.pTrack->UpdatePathAndIndex(obj.predTrajectories.at(f.new_index), f.new_index);
+
+			bool bFound = false;
+			for(unsigned int k=0; k < check_list.size(); k++)
```
It's better to use standard functions. It seems that this loop can be replaced by std::find.

hatem-darweesh replied:
personal preference

kfunaoka replied:
@hatem-darweesh NP. But std::find can reduce the number of lines 1/9 (one ninth) where we develop, review, and maintain. If there are ten loops finding an element, it can reduce 81 (= 90 - 9) lines in the Autoware code base. If you like it, please use it :)

- The number of lines of this loop is 9 (from Line 688 to 695).
- The number of lines of std::find is 1.

kfunaoka commented on ros/src/computing/perception/detection/lidar_tracker/packages/lidar_kf_contour_track/nodes/lidar_kf_contour_track/PolygonGenerator.cpp:
```diff
@@ -0,0 +1,113 @@
+
+std::vector<PlannerHNS::GPSPoint> PolygonGenerator::EstimateClusterPolygon(const pcl::PointCloud<pcl::PointXYZ>& cluster, const PlannerHNS::GPSPoint& original_centroid, PlannerHNS::GPSPoint& new_centroid, const double& polygon_resolution)
+{
+	for(unsigned int i=0; i < m_Quarters.size(); i++)
```
You can use range-for.
https://en.cppreference.com/w/cpp/language/range-for

hatem-darweesh replied:
personal preference

kfunaoka replied:
Same as https://github.com/CPFL/Autoware/pull/1400#discussion_r210462246

kfunaoka commented on ros/src/computing/planning/common/lib/openplanner/op_planner/include/op_planner/PlanningHelpers.h:
```diff
@@ -1,14 +1,14 @@
 #define PLANNINGHELPERS_H_
 
-#include <math.h>
+//#include <math.h>
```
Should be deleted. Unnecessary comments will be junk.

kfunaoka commented on ros/src/computing/planning/common/lib/openplanner/op_planner/src/BehaviorPrediction.cpp:
```diff
@@ -0,0 +1,762 @@
+	}
+}
+
+int BehaviorPrediction::FromIndicatorToNumber(const PlannerHNS::LIGHT_INDICATOR& indi)
```
Why are FromIndicatorToNumber and FromNumbertoIndicator necessary? Is it impossible to use enum value itself?

hatem-darweesh replied:
the compiler give me an error casting from int to the enumerator, so I create a function to convert from one to another and the vice versa , just to be consistent

kfunaoka replied:
@hatem-darweesh Is it just a design problem or an unavoidable problem?  Basically, it seems that it is not a good idea to have the same type of data (Particle::indicator and the data returned from BehaviorPrediction::FromNumbertoIndicator) with different types (int and lannerHNS::LIGHT_INDICATOR, respectively).

kfunaoka replied:
I see. Particle::indicator is defined as int in autoware_msgs::DetectedObject. It is difficult to avoid.

kfunaoka commented on ros/src/computing/planning/common/lib/openplanner/op_planner/src/BehaviorPrediction.cpp:
```diff
@@ -0,0 +1,762 @@
+double BehaviorPrediction::CalcIndicatorWeight(PlannerHNS::LIGHT_INDICATOR p_ind, PlannerHNS::LIGHT_INDICATOR obj_ind)
+{
+	if((obj_ind == PlannerHNS::INDICATOR_LEFT || obj_ind == PlannerHNS::INDICATOR_RIGHT || obj_ind == PlannerHNS::INDICATOR_NONE) && p_ind == obj_ind)
+		return 0.99;
```
Please remove magic numbers as much as possible.

hatem-darweesh replied:
This part (Prediction) is under development (research work) , all numbers will be removed or added as parameters soon.

kfunaoka reviewed (commented):
@hatem-darweesh I've added some comments, but it is difficult to review whole because the diff is very wide range. Would you list reasons why these changes are necessary for each items with comparing with the old implementation?

- OpenPlanner - Global planner
- OpenPlanner - Local Planner
- OpenPlanner - Utilities
- OpenPlanner - Simulator
- lidar_kf_contour_track

kfunaoka commented on ros/src/computing/planning/mission/packages/op_global_planner/package.xml:
```diff
@@ -0,0 +1,78 @@
+  <!--   <run_depend>message_runtime</run_depend> -->
+  <!-- Use test_depend for packages you need only for testing: -->
+  <!--   <test_depend>gtest</test_depend> -->
+  <buildtool_depend>catkin</buildtool_depend>
```
Please add buildtool_depend of autoware_build_flags.
https://github.com/CPFL/Autoware/wiki/Quality-Control

kfunaoka commented on ros/src/computing/planning/mission/packages/op_global_planner/package.xml:
```diff
@@ -0,0 +1,78 @@
+  <version>0.0.1</version>
+  <description>The OpenPlanner Global Planning package</description>
+
+  <!-- One maintainer tag required, multiple allowed, one person per tag --> 
```
Please delete unnecessary comments.

kfunaoka commented on ros/src/computing/planning/mission/packages/op_global_planner/CMakeLists.txt:
```diff
@@ -0,0 +1,52 @@
+project(op_global_planner)
+
+find_package(catkin REQUIRED COMPONENTS
+  geometry_msgs
```
Please add autoware_build_flags.
https://github.com/CPFL/Autoware/wiki/Quality-Control

hatem-darweesh committed:
Merge branch 'develop' into feature/OpenPlanner-dev

hatem-darweesh committed:
Update BehaviorPrediction.h

pass pointer as "const" for function "DeleteFromList" in both (BehaviorPrediction and LLP) classes

hatem-darweesh committed:
Update PlanningHelpers.h

Delete useless comment for including math.h

kfunaoka reviewed (dismissed):
@hatem-darweesh Though I've made a lot of comments, my mandatory requests now are:

- Inserting license information into *.{h,cpp} files as mentioned at https://github.com/CPFL/Autoware/pull/1400#discussion_r210443001.
- Inserting autoware_build_flags into package.xml and CMakeLists.txt as shown at https://github.com/CPFL/Autoware/pull/1400#pullrequestreview-146381373
- Updating the headline of this pull request to show what are different from before and why they are necessary. Your slide and descriptions will help us to understand your work.

hatem-darweesh committed:
Support old behavior of insert static object for obstacle avoidance testing
Only one simulated car available in the runtime manager
update for copywrite note
insert autoware_build_flags to new nodes

hatem-darweesh committed:
Merge branch 'develop' into feature/OpenPlanner-dev

hatem-darweesh committed:
Merge develop into OpenPlanner-dev

kfunaoka commented on ros/src/computing/perception/detection/lidar_tracker/packages/lidar_kf_contour_track/CMakeLists.txt:
```diff
@@ -8,10 +8,10 @@ find_package(catkin REQUIRED COMPONENTS
-
-set(CMAKE_CXX_FLAGS "-O2 -Wall ${CMAKE_CXX_FLAGS}")
+find_package(OpenCV REQUIRED)
 
```
```diff
- set(CMAKE_CXX_FLAGS "-O2 -Wall ${CMAKE_CXX_FLAGS}")
```

Should we revert this line because you said

> The other flags are needed only locally for debugging and performance test so I don't want to push them to the server

hatem-darweesh replied:
@kfunaoka  I will add to different PR if necessary

kfunaoka reviewed (commented):
@hatem-darweesh I've confirmed that all my mandatory request are OK. I might find an item that you want to change.
